src/net.rs

```diff
@@ -199,19 +207,336 @@ impl UdpCodec for LineCodec {
-    let port = if let Some(port) = opt_port { port } else { PORT };
+    let host = if let Some(host) = opt_host { host } else { DEFAULT_HOST };
+    let port = if let Some(port) = opt_port { port } else { DEFAULT_PORT };
     let addr: SocketAddr = format!("{}:{}", host, port).parse()?;
     let sock = UdpSocket::bind(&addr, &handle).expect("failed to bind socket");
     Ok(sock)
 }
 
-//XXX other functions
-
-pub const HOST: &str = "0.0.0.0";
-pub const PORT: u16 = 12345;
-
 #[allow(dead_code)]
 pub fn get_version() -> result::Result<Version, SemVerError> {
     Version::parse(VERSION)
 }
+
+pub struct NetworkStatistics {
+    packets_tx_failed: u64,
+    packets_tx_success: u64,
+    keep_alive_tx_failed: u64,
+    keep_alive_tx_success: u64,
+}
+
+impl NetworkStatistics {
+    fn new() -> Self {
+        NetworkStatistics {
+            packets_tx_success: 0,
+            packets_tx_failed: 0,
+            keep_alive_tx_failed: 0,
+            keep_alive_tx_success: 0
+        }
+    }
+
+    pub fn inc_tx_packets_failed(&mut self) {
+        self.packets_tx_failed += 1;
+    }
+
+    pub fn inc_tx_packets_success(&mut self) {
+        self.packets_tx_success += 1;
+    }
+
+    pub fn inc_tx_keep_alive_failed(&mut self) {
+        self.keep_alive_tx_failed += 1;
+    }
+
+    pub fn inc_tx_keep_alive_success(&mut self) {
+        self.keep_alive_tx_success += 1;
+    }
+}
+
+pub struct NetworkManager {
+    pub statistics:     NetworkStatistics,
+    pub tx_packets:     VecDeque<Packet>, // Back == Newest, Front == Oldest
+    rx_packets:     VecDeque<Packet>, // Back == Newest, Front == Oldest
+}
+
+impl NetworkManager {
+    pub fn new() -> Self {
+        NetworkManager {
+            statistics: NetworkStatistics::new(),
+            tx_packets:  VecDeque::<Packet>::with_capacity(PACKET_HISTORY_SIZE),
+            rx_packets:  VecDeque::<Packet>::with_capacity(PACKET_HISTORY_SIZE),
+        }
+    }
+
+    pub fn head_of_tx_packet_queue(&self) -> Option<&Packet> {
+        self.tx_packets.back()
+    }
+
+    /// The TX Packet queue must only contain Request packets.
+    fn newest_tx_packet_seq_num(&self) -> Option<u64> {
+        let opt_newest_packet = self.head_of_tx_packet_queue();
+
+        if opt_newest_packet.is_some() {
+            let newest_packet = opt_newest_packet.unwrap();
+            if let Packet::Request{ sequence: newest_sequence, response_ack: _, cookie: _, action: _ } = newest_packet {
+                Some(*newest_sequence)
+            } else { panic!("Found something other than a `Request` packet in the buffer: {:?}", newest_packet) }
+                    // Somehow not a Request packet. Panic during development XXX
+        } else { None }
+                // Queue is empty
+    }
+
+    /// This will keep the specified queue under the PACKET_HISTORY_SIZE limit.
+    /// The TX queue needs to ensure a spot is open if we're at capacity.
+    fn discard_older_packets(&mut self, is_tx_queue: bool) {
```

Comment:
As it stands, you couldn't do that because there would be a battle to borrow `self` and the queue in question at the same time. I *think* I'd have to add in explicit lifetimes to get around that and it seems like a lot of work for a little gain. The other option is to remove this from `NetworkManager`, and thus `self`, but tbh I don't want to move this into a standalone function since that doesn't make sense to me. If you're worried about at-a-glance readability, I can define two private boolean constants to help with that.
```Rust
const DISCARD_TX_QUEUE: boolean = true;
const DISCARD_RX_QUEUE: boolean = false;

...
network_manager.discard_older_packets(DISCARD_TX_QUEUE);
```

Reply:
(Going to try and correct this using traits in `reliability part 2`) 😃

Reply:
I'd prefer an `enum Queue {Tx, Rx}`. That's more Rusty :)